### ISSUE TITLE
Change Request: Add option to report on unused/redundant inline configs / configuration comments

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [eslint/eslint#18230]
### ESLint version

8.57.0

### What problem do you want to solve?

Right now, nothing in ESLint core stops [inline configs (configuration comments)](https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments) from redundantly repeating the same configuration _option_ and/or _severity_ as already exists. This is similar to [unused disable directives](https://eslint.org/docs/latest/use/configure/rules#report-unused-eslint-disable-comments): they take up space and can be misleading.

* [Playground of an inline config setting the same severity as the config file](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC1leHByZXNzaW9uczogXCJlcnJvclwiICovXG5cIuKdjCBkb2VzIG5vdGhpbmc6IGV4aXN0aW5nIHNldmVyaXR5IHZzLiBmaWxlXCJcblxuLyogZXNsaW50IG5vLXVudXNlZC1leHByZXNzaW9uczogXCJ3YXJuXCIgKi9cblwi4pyFIGRvZXMgc29tZXRoaW5nOiBuZXcgc2V2ZXJpdHkgdnMuIGNvbW1lbnRcIlxuXG4vKiBlc2xpbnQgbm8tdW51c2VkLWV4cHJlc3Npb25zOiBcImVycm9yXCIgKi9cblwi4pyFIGRvZXMgc29tZXRoaW5nOiBuZXcgc2V2ZXJpdHkgdnMuIGNvbW1lbnRcIlxuIiwib3B0aW9ucyI6eyJlbnYiOnt9LCJydWxlcyI6e30sInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fSwiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoibW9kdWxlIn19fQ==)
* [Playground of an inline config setting the same options as the config file](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC1leHByZXNzaW9uczogW1wiZXJyb3JcIiwgeyBcImFsbG93U2hvcnRDaXJjdWl0XCI6IHRydWUgfV0gKi9cblwi4p2MIGRvZXMgbm90aGluZzogZXhpc3Rpbmcgc2V2ZXJpdHkgYW5kIG9wdGlvbnMgdnMuIGZpbGVcIlxuXG4vKiBlc2xpbnQgbm8tdW51c2VkLWV4cHJlc3Npb25zOiBbXCJlcnJvclwiLCB7IFwiYWxsb3dTaG9ydENpcmN1aXRcIjogZmFsc2UgfV0gKi9cblwi4pyFIGRvZXMgc29tZXRoaW5nOiBleGlzdGluZyBzZXZlcml0eSBhbmQgb3B0aW9ucyB2cy4gY29tbWVudFwiXG4iLCJvcHRpb25zIjp7ImVudiI6e30sInJ1bGVzIjp7Im5vLXVudXNlZC1leHByZXNzaW9ucyI6WyJlcnJvciIseyJhbGxvd1Nob3J0Q2lyY3VpdCI6dHJ1ZSwiYWxsb3dUZXJuYXJ5IjpmYWxzZSwiYWxsb3dUYWdnZWRUZW1wbGF0ZXMiOmZhbHNlLCJlbmZvcmNlRm9ySlNYIjpmYWxzZX1dfSwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9LCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUifX19)

### What do you think is the correct solution?

The most straightforward change I can think of would be to add a new `--report-unused-inline-configs` / `reportUnusedInlineConfigs` (per the feature's primary naming in #18187) or `--report-unused-config-comments` / `reportUnusedConfigComments` (per the common alternate naming).

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

Inspired by #18218, thanks @inga-lovinde 😄 

Edit: https://github.com/eslint/rfcs/pull/121 was filed and accepted for this.

**Comment by nzakas** [comment]
This seems like a good idea that is worth exploring in an RFC. Keep in mind, that going forward any new options like this need to work with the [language plugins design](https://github.com/eslint/rfcs/blob/main/designs/2022-languages/README.md), which makes things a bit more difficult because we have to take into account how other languages might do inline configs.

**Comment by fasttime** [comment]
If I'm not misunderstanding the problem, this has been already addressed in #18157. In v9, redundant `eslint` config comments will produce an error ([**see repro**](https://stackblitz.com/edit/stackblitz-starters-smtst8)).

**Comment by nzakas** [comment]
@fasttime ah good point! @JoshuaKGoldberg do you agree?

**Comment by JoshuaKGoldberg** [comment]
Ah, I'd missed that, thank you. I think there's only partial overlap though. If I've read it all correctly, #18132 was purely around comments in a single file; this one is both for that _(overlap)_ and comparing comments to the backing config file.

If a rule is configured in a comment the same way as it was in a config file, #18157's changes won't report anything:

```js
// eslint.config.js
export default [
    {
        rules: {
            "prefer-const": "error"
        }
    }
];
```

```js
// index.js
/* eslint prefer-const: "error" */
export const a = 0;
```

**Comment by nzakas** [comment]
Got it, thanks for explaining. In that case, we're back on the RFC track for this.

**Comment by inga-lovinde** [comment]
I think this is still relevant

**Comment by nzakas** [comment]
@inga-lovinde I agree, it's still relevant, but we need an [RFC](https://github.com/eslint/rfcs) to proceed.

**Comment by Rec0iL99** [comment]
Waiting for an RFC.

**Comment by JoshuaKGoldberg** [comment]
I am, at present, working on the RFC.

**Comment by JoshuaKGoldberg** [comment]
RFC posted: https://github.com/eslint/rfcs/pull/121

Note that the proposed new option is visually similar to the existing `--report-unused-disable-directives(-severity)` / `linterOptions.reportUnusedDisableDirectives`. I filed two issues around smoothing out the existing areas:

* #18665
* #18666

**Comment by ljharb** [comment]
that wasn't 7 days, that was 0 days

**Comment by nzakas** [comment]
RFC is open, so not stale.

@ljharb it's seven days from today, not seven days since the last comment. :)

**Comment by JoshuaKGoldberg** [comment]
I re-opened #121.

**Comment by nzakas** [comment]
@JoshuaKGoldberg that's a very old issue...is it possible you mistyped?

**Comment by JoshuaKGoldberg** [comment]
🤦 yes, wrong repository - thanks.

**Comment by JoshuaKGoldberg** [comment]
Just confirming the process: this is ready to be implemented, right?

**Comment by fasttime** [comment]
> Just confirming the process: this is ready to be implemented, right?

Yes, a PR is welcome 👍
